File: JavaScript/tower_of_hanoi_recursive.js
// Tower of Hanoi - Recursive Solution

// Problem Statement :
// The Towers of Hanoi has 3 pegs with multiple disks of different sizes.
// The goal of the puzzle is to move all the disks from the first peg to the third peg according to the following rules :
// Only one disk can be moved at a time.
// You can only move the top disc in a stack.
// No disk may be placed on top of a smaller disk.

// Javascript Code :

function towerOfHanoi(disksnumber, startpeg, endpeg, middlepeg) {
  if (disksnumber >= 1) {

    // Move a tower of disksnumber-1 to the middle peg, using the end peg.
    towerOfHanoi(disksnumber - 1, startpeg, middlepeg, endpeg);

    // Move the remaining disks to the end peg.
    console.log('Move disk from Tower ', startpeg, ' to Tower ', endpeg);

    // Move the tower of `disksnumber-1` from the `middle peg` to the `end peg` using the `start peg`.        
    towerOfHanoi(disksnumber - 1, middlepeg, endpeg, startpeg);
  }
  
  return;
}

towerOfHanoi(3, "A", "C", "B");


/*
---Details---

---Sample Inputs and Outputs---

No input required.
Output:
Move disk from Tower  A  to Tower  C
Move disk from Tower  A  to Tower  B
Move disk from Tower  C  to Tower  B
Move disk from Tower  A  to Tower  C
Move disk from Tower  B  to Tower  A
Move disk from Tower  B  to Tower  C
Move disk from Tower  A  to Tower  C

---Time Complexity---
Time Complexity is O(n^2)

---Space Complexity---
Space Complexity is O(n)
*/

File: Go/towerOfHanoi_Recursive.go
package main
import (
    "fmt"
     "os"
    "strconv"
)

func towerOfHanoi(n int, a, b, c string){
    if n == 1{
        fmt.Printf("move from %s to %s\n", a, c)
        return
    }

    towerOfHanoi(n-1, a, c, b)
    fmt.Printf("move from %s to %s\n", a,c)
    towerOfHanoi(n-1, b, a, c)
}

func main(){
    var n int = 3
    var err error
    if len(os.Args) == 2{
        n, err = strconv.Atoi(os.Args[1])
        if err != nil {
            fmt.Printf("%s\n", err.Error())
            os.Exit(-1)
        }
    }

    towerOfHanoi(n, "A", "B", "C")
}

// Sample Input 
// 	3, A, B, C
// Sample Output
// move from A to C
// move from A to B
// move from C to B
// move from A to C
// move from B to A
// move from B to C
// move from A to C

// Time Complexity - O(2^n)
// Space Complexity - O(n)

